rig/machine_control/regions.py

```diff
@@ -144,75 +117,102 @@ def __init__(self, base_x=0, base_y=0, level=0):
+        # which subregions this core should be filled.
+        self.locally_selected = array.array('H', (0x0 for _ in range(18)))
 
-        # And possibly contains subregions
+        # If this is a coarser region tree then we also maintain a subtree of
+        # fixed size.
         if level < 3:
             self.subregions = [None] * 16
 
-    def get_regions(self):
-        """Generate a set of integer region representations.
+    def get_regions_and_coremasks(self):
+        """Generate a set of ordered paired region and core mask representations.
 
-        Returns
-        -------
-        generator
-            Generator which yields 32-bit region codes as might be generated by
-            :py:func:`.get_region_for_chip`.
+        .. note::
+            The region and core masks are ordered such that ``(region << 32) |
+            core_mask`` is monotonically increasing. Consequently region and
+            core masks generated by this method can be used with SCAMP's
+            Flood-Fill Core Select (FFSC) method.
+
+        Yields
+        ------
+        (region, core mask)
+            Pair of integers which represent a region of a SpiNNaker machine
+            and a core mask of selected cores within that region.
         """
         region_code = ((self.base_x << 24) | (self.base_y << 16) |
                        (self.level << 16))
 
-        # Build up the returned set of regions
-        if self.locally_selected != set():
-            elements = 0x0000
-            for e in self.locally_selected:
-                elements |= 1 << e
-            yield (region_code | elements)
+        # Generate core masks for any regions which are selected at this level
+        # Create a mapping from subregion mask to core numbers
+        subregions_cores = collections.defaultdict(lambda: 0x0)
+        for core, subregions in enumerate(self.locally_selected):
+            if subregions:  # If any subregions are selected on this level
+                subregions_cores[subregions] |= 1 << core
 
-        # Include subregions if they exist
-        if self.level < 3:
-            for i, sr in enumerate(self.subregions):
-                if i not in self.locally_selected and sr is not None:
-                    for r in sr.get_regions():
-                        yield r
+        # Order the locally selected items and then yield them
+        sorted_subregion_coremasks = sorted(list(subregions_cores.items()),
+                                            key=lambda x: (x[0] << 32) | x[1])
+        for (subregions, coremask) in sorted_subregion_coremasks:
+            yield (region_code | subregions), coremask
 
-    def add_coordinate(self, x, y):
-        """Add a new coordinate to the region tree.
+        if self.level < 3:
+            # Iterate through the subregions and recurse, we iterate through in
+            # the order which ensures that anything we yield is in increasing
+            # order.
```

Comment:
It expects the combination `(region << 18) | coremask` in increasing orders... so yeah, low-numbered levels go first because they are the next significant field after y and x.